In [4]:
import Clustering;

In [28]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [29]:
# Let's port over some dataframes into this notebook
data = Clustering.imp_data

##### The aim is to conduct statistical tests between each cluster

In [35]:
# Add cluster labels to the data
cluster_labels = pd.DataFrame(Clustering.pca_cluster_labels)
data['Cluster_Labels'] = cluster_labels.replace({0: 'Cluster 1', 1: 'Cluster 2', 2: 'Cluster 3'})
data

BALANCE  BALANCE_FREQUENCY  PURCHASES  ONEOFF_PURCHASES  \
0       40.900749           0.818182      95.40              0.00   
1     3202.467416           0.909091       0.00              0.00   
2     2495.148862           1.000000     773.17            773.17   
3     1666.670542           0.636364    1499.00           1499.00   
4      817.714335           1.000000      16.00             16.00   
...           ...                ...        ...               ...   
8945    28.493517           1.000000     291.12              0.00   
8946    19.183215           1.000000     300.00              0.00   
8947    23.398673           0.833333     144.40              0.00   
8948    13.457564           0.833333       0.00              0.00   
8949   372.708075           0.666667    1093.25           1093.25   

      INSTALLMENTS_PURCHASES  CASH_ADVANCE  PURCHASES_FREQUENCY  \
0                      95.40      0.000000             0.166667   
1                       0.00   6442.945483             0.000000   
2                       0.00      0.000000             1.000000   
3                       0.00    205.788017             0.083333   
4                       0.00      0.000000             0.083333   
...                      ...           ...                  ...   
8945                  291.12      0.000000             1.000000   
8946                  300.00      0.000000             1.000000   
8947                  144.40      0.000000             0.833333   
8948                    0.00     36.558778             0.000000   
8949                    0.00    127.040008             0.666667   

      ONEOFF_PURCHASES_FREQUENCY  PURCHASES_INSTALLMENTS_FREQUENCY  \
0                       0.000000                          0.083333   
1                       0.000000                          0.000000   
2                       1.000000                          0.000000   
3                       0.083333                          0.000000   
4                       0.083333                          0.000000   
...                          ...                               ...   
8945                    0.000000                          0.833333   
8946                    0.000000                          0.833333   
8947                    0.000000                          0.666667   
8948                    0.000000                          0.000000   
8949                    0.666667                          0.000000   

      CASH_ADVANCE_FREQUENCY  CASH_ADVANCE_TRX  PURCHASES_TRX  CREDIT_LIMIT  \
0                   0.000000               0.0            2.0        1000.0   
1                   0.250000               4.0            0.0        7000.0   
2                   0.000000               0.0           12.0        7500.0   
3                   0.083333               1.0            1.0        7500.0   
4                   0.000000               0.0            1.0        1200.0   
...                      ...               ...            ...           ...   
8945                0.000000               0.0            6.0        1000.0   
8946                0.000000               0.0            6.0        1000.0   
8947                0.000000               0.0            5.0        1000.0   
8948                0.166667               2.0            0.0         500.0   
8949                0.333333               2.0           23.0        1200.0   

         PAYMENTS  MINIMUM_PAYMENTS  PRC_FULL_PAYMENT  TENURE  \
0      201.802084        139.509787          0.000000    12.0   
1     4103.032597       1072.340217          0.222222    12.0   
2      622.066742        627.284787          0.000000    12.0   
3        0.000000        379.829982          0.000000    12.0   
4      678.334763        244.791237          0.000000    12.0   
...           ...               ...               ...     ...   
8945   325.594462         48.886365          0.500000     6.0   
8946   275.861322        156.060050          0.000000     6.0   
8

In [50]:
data.describe().T

count         mean          std        min  \
BALANCE                           8950.0  1564.474828  2081.531879   0.000000   
BALANCE_FREQUENCY                 8950.0     0.877271     0.236904   0.000000   
PURCHASES                         8950.0  1003.204834  2136.634782   0.000000   
ONEOFF_PURCHASES                  8950.0   592.437371  1659.887917   0.000000   
INSTALLMENTS_PURCHASES            8950.0   411.067645   904.338115   0.000000   
CASH_ADVANCE                      8950.0   978.871112  2097.163877   0.000000   
PURCHASES_FREQUENCY               8950.0     0.490351     0.401371   0.000000   
ONEOFF_PURCHASES_FREQUENCY        8950.0     0.202458     0.298336   0.000000   
PURCHASES_INSTALLMENTS_FREQUENCY  8950.0     0.364437     0.397448   0.000000   
CASH_ADVANCE_FREQUENCY            8950.0     0.135144     0.200121   0.000000   
CASH_ADVANCE_TRX                  8950.0     3.248827     6.824647   0.000000   
PURCHASES_TRX                     8950.0    14.709832    24.857649   0.000000   
CREDIT_LIMIT                      8950.0  4494.179679  3638.701915  50.000000   
PAYMENTS                          8950.0  1733.143852  2895.063757   0.000000   
MINIMUM_PAYMENTS                  8950.0   852.716172  2347.654983   0.019163   
PRC_FULL_PAYMENT                  8950.0     0.153715     0.292499   0.000000   
TENURE                            8950.0    11.517318     1.338331   6.000000   
Credit Card Utilization Ratio     8950.0     0.388884     0.389721   0.000000   

                                          25%          50%          75%  \
BALANCE                            128.281915   873.385231  2054.140036   
BALANCE_FREQUENCY                    0.888889     1.000000     1.000000   
PURCHASES                           39.635000   361.280000  1110.130000   
ONEOFF_PURCHASES                     0.000000    38.000000   577.405000   
INSTALLMENTS_PURCHASES               0.000000    89.000000   468.637500   
CASH_ADVANCE                         0.000000     0.000000  1113.821139   
PURCHASES_FREQUENCY                  0.083333     0.500000     0.916667   
ONEOFF_PURCHASES_FREQUENCY           0.000000     0.083333     0.300000   
PURCHASES_INSTALLMENTS_FREQUENCY     0.000000     0.166667     0.750000   
CASH_ADVANCE_FREQUENCY               0.000000     0.000000     0.222222   
CASH_ADVANCE_TRX                     0.000000     0.000000     4.000000   
PURCHASES_TRX                        1.000000     7.000000    17.000000   
CREDIT_LIMIT                      1600.000000  3000.000000  6500.000000   
PAYMENTS                           383.276166   856.901546  1901.134317   
MINIMUM_PAYMENTS                   166.310858   300.956126   804.414788   
PRC_FULL_PAYMENT                     0.000000     0.000000     0.142857   
TENURE                              12.000000    12.000000    12.000000   
Credit Card Utilization Ratio        0.041494     0.302720     0.717571   

                                           max  
BALANCE                           19043.138560  
BALANCE_FREQUENCY                     1.000000  
PURCHASES                         49039.570000  
ONEOFF_PURCHASES                  40761.250000  
INSTALLMENTS_PURCHASES            22500.000000  
CASH_ADVANCE                      47137.211760  
PURCHASES_FREQUENCY                   1.000000  
ONEOFF_PURCHASES_FREQUENCY            1.000000  
PURCHASES_INSTALLMENTS_FREQUENCY      1.000000  
CASH_ADVANCE_FREQUENCY                1.500000  
CASH_ADVANCE_TRX                    123.000000  
PURCHASES_TRX                       358.000000  
CREDIT_LIMIT                      30000.000000  
PAYMENTS                          50721.483360  
MINIMUM_PAYMENTS                  76406.207520  
PRC_FULL_PAYMENT                      1.000000  
TENURE                               12.000000  
Credit Card Utilization Ratio        15.909951

In [39]:
# Next: segment data into 3 dataframe subsets

Cluster1_data = data[data['Cluster_Labels'] == "Cluster 1"]
Cluster2_data = data[data['Cluster_Labels'] == "Cluster 2"]
Cluster3_data = data[data['Cluster_Labels'] == "Cluster 3"]

In [46]:
print("There are {} observations in Cluster 1".format(len(Cluster1_data)))
print("There are {} observations in Cluster 2".format(len(Cluster2_data)))
print("There are {} observations in Cluster 3".format(len(Cluster3_data)))

There are 4721 observations in Cluster 1
There are 2804 observations in Cluster 2
There are 1425 observations in Cluster 3


In [65]:
from scipy import stats

# perform an ANOVA test on feature means between all clusters on 
Credit_limit = 'CREDIT_LIMIT'
statistic, p_value = stats.f_oneway(Cluster1_data[Credit_limit], Cluster2_data[Credit_limit], Cluster3_data[Credit_limit])
print(f"F-value: {statistic}")
print(f"p-value: {p_value}")

F-value: 349.9262895344267
p-value: 4.7846195737917477e-147


##### The high F-value tells us that there exist statistically significant differences in the mean credit limits of cardholders between clusters. Let's investigate to see which clusters have the highest and lowest mean credit limits:

In [75]:
print(f'Cluster 1: \n', Cluster1_data["CREDIT_LIMIT"].describe(), '\n')
print(f'Cluster 2: \n', Cluster2_data["CREDIT_LIMIT"].describe(), '\n')
print(f'Cluster 3: \n', Cluster3_data["CREDIT_LIMIT"].describe())

Cluster 1: 
 count     4721.000000
mean      4139.798322
std       3339.599404
min         50.000000
25%       1500.000000
50%       3000.000000
75%       6000.000000
max      23000.000000
Name: CREDIT_LIMIT, dtype: float64 

Cluster 2: 
 count     2804.000000
mean      3950.543145
std       3234.497051
min        300.000000
25%       1500.000000
50%       3000.000000
75%       5500.000000
max      21500.000000
Name: CREDIT_LIMIT, dtype: float64 

Cluster 3: 
 count     1425.000000
mean      6737.962998
std       4428.232982
min        300.000000
25%       3200.000000
50%       6000.000000
75%       9000.000000
max      30000.000000
Name: CREDIT_LIMIT, dtype: float64


##### The mean cardholder credit limit is highest in Cluster 3 at $6737.96 compared to the lowest in Cluster 2 at $3950.54

In [79]:
# Let's run a t-test between Cluster 1 and Cluster 2:

t_stat, t_test_p_value = stats.ttest_ind(Cluster1_data[Credit_limit], Cluster2_data[Credit_limit])
t_test_p_value

0.016205731729706286

##### At a p=0.016, we can be sure that the pairwise difference between Cluster 1 and Cluster 2 is statistically significant - now, I want to run a Tukey HSD post-hoc test to test all possible pairwise differences

In [84]:
from statsmodels.formula.api import ols
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Combine all the datasets into one
combined_data = pd.concat([Cluster1_data, Cluster2_data, Cluster3_data])

# Perform ANOVA
model = ols('CREDIT_LIMIT ~ Cluster_Labels', data=combined_data).fit()
anova_table = sm.stats.anova_lm(model)

# Perform Tukey's HSD post hoc test
posthoc = pairwise_tukeyhsd(combined_data['CREDIT_LIMIT'], combined_data['Cluster_Labels'])
print(posthoc)

      Multiple Comparison of Means - Tukey HSD, FWER=0.05      
  group1    group2   meandiff p-adj    lower     upper   reject
---------------------------------------------------------------
Cluster 1 Cluster 2 -189.2552 0.0609 -385.1228    6.6125  False
Cluster 1 Cluster 3 2598.1647    0.0 2349.8586 2846.4707   True
Cluster 2 Cluster 3 2787.4199    0.0 2520.1574 3054.6823   True
---------------------------------------------------------------


##### We can see that the Tukey HSD test cannot say that the difference in means between the first two clusters is statistically significant, ergo we can't yet say that there exist significant differences in credit limits between Cluster 1 and Cluster 2. However, Tukey does confirm the difference between Cluster 3 and the other two clusters, suggesting that Cluster 3 is comprised of individuals with higher annual incomes and credit scores given their higher spending power and allocated credit limit.